genesis/process/intermediate/deployLibrarySC.go

```diff
@@ -0,0 +1,149 @@
+package intermediate
+
+import (
+	"bytes"
+	"encoding/hex"
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/genesis"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+// ArgDeployProcessor is the argument used to create a deployLibrarySC instance
+type ArgDeployLibrarySC struct {
+	Executor         genesis.TxExecutionProcessor
+	PubkeyConv       core.PubkeyConverter
+	BlockchainHook   process.BlockChainHookHandler
+	ShardCoordinator sharding.Coordinator
+}
+
+type deployLibrarySC struct {
+	*baseDeploy
+	shardCoordinator sharding.Coordinator
+}
+
+// NewDeployLibrarySC returns a new instance of deploy library SC able to deploy library SC that needs to be present
+// on all shards - same contract is deployed core.MaxShards time with addresses which end with all possibilities of the
+// last 2 bytes
+func NewDeployLibrarySC(arg ArgDeployLibrarySC) (*deployLibrarySC, error) {
+	if check.IfNil(arg.Executor) {
+		return nil, genesis.ErrNilTxExecutionProcessor
+	}
+	if check.IfNil(arg.PubkeyConv) {
+		return nil, genesis.ErrNilPubkeyConverter
+	}
+	if check.IfNil(arg.BlockchainHook) {
+		return nil, process.ErrNilBlockChainHook
+	}
+	if check.IfNil(arg.ShardCoordinator) {
+		return nil, genesis.ErrNilShardCoordinator
+	}
+
+	base := &baseDeploy{
+		TxExecutionProcessor: arg.Executor,
+		pubkeyConv:           arg.PubkeyConv,
+		blockchainHook:       arg.BlockchainHook,
+		emptyAddress:         make([]byte, arg.PubkeyConv.Len()),
+		getScCodeAsHex:       getSCCodeAsHex,
+	}
+
+	dp := &deployLibrarySC{
+		baseDeploy:       base,
+		shardCoordinator: arg.ShardCoordinator,
+	}
+
+	return dp, nil
+}
+
+func GenerateInitialPublicKeys(
+	baseAddress []byte,
+	shardCoordinator sharding.Coordinator,
+	allShards bool,
+) [][]byte {
+	addressLen := len(baseAddress)
+
+	newAddresses := make([][]byte, 0)
+	i := uint8(0)
+	for ; i < core.MaxShardNumber-1; i++ {
+		shardInBytes := []byte{0, i}
+		tmpAddress := string(baseAddress[:(addressLen-core.ShardIdentiferLen)]) + string(shardInBytes)
+
+		newShardID := shardCoordinator.ComputeId([]byte(tmpAddress))
+		if !allShards && newShardID != shardCoordinator.SelfId() {
```

Comment:
Perhaps convert this into a positive condition, remove the `continue` and do `newAddresses = ...`? (easier to follow).

Is this maybe:

```
if allShards || newShardID == shardCoordinator.SelfId() {
	newAddresses = append(newAddresses, []byte(tmpAddress))
}
```

If so, adjust the end condition of the loop (without -1) and remove the steps after the loop (I feel I am missing something)?

Reply:
yes :) uint8 and would go into an infinite loop.